grpc/runtime/src/main/scala/io/buoyant/grpc/runtime/DecodingStream.scala

```diff
@@ -0,0 +1,326 @@
+   * being constructed, it is paired with a _releasable_ function that
+   * frees the underlying H2 frames as needed.
+   */
+  private trait Releaser {
+
+    /**
+     * Add the given H2 frame to be tracked and released.
+     */
+    def track(frame: h2.Frame): Releaser
+
+    /**
+     * Mark `n` bytes as consumed.
+     */
+    def consume(n: Int): Releaser
+
+    /**
+     * Reap all consumed frames into a releasable function, and return
+     * the updated state afterwards.
+     */
+    def releasable(): (Releaser, Releaser.Func)
+  }
+
+  private object Releaser {
+    type Func = () => Future[Unit]
+    val NopFunc: Func = () => Future.Unit
+
+    object Underflow
+      extends Exception("released too many bytes")
+      with NoStackTrace
+
+    object Nil extends Releaser {
+      override def toString = "Releaser.Nil"
+      override def track(f: h2.Frame): Releaser = mk(f)
+
+      override def consume(n: Int): Releaser =
+        if (n == 0) this
+        else throw Underflow
+
+      override def releasable(): (Releaser, Func) = _releasable
+      private[this] val _releasable = (this, NopFunc)
+    }
+
+    private[this] case class FrameReleaser(
+      consumed: Int,
+      total: Int,
+      release: () => Future[Unit],
+      tail: Releaser
+    ) extends Releaser {
+      require(0 <= consumed && consumed <= total)
+
+      def remaining: Int = total - consumed
+
+      override def toString: String =
+        s"Releaser.Frame($consumed, $total, $tail)"
+
+      override def track(f: h2.Frame): Releaser =
+        copy(tail = tail.track(f))
+
+      override def consume(n: Int): Releaser =
+        if (n == 0) this
+        else if (n <= remaining) copy(consumed = consumed + n)
+        else {
+          // Leave a 0-length releasable, so that `release` is chained
+          // into the tail `releasable` call.
+          copy(
+            consumed = total,
+            tail = tail.consume(n - remaining)
+          )
+        }
+
+      override def releasable(): (Releaser, Func) =
+        if (consumed < total) {
+          // This frame isn't yet releasable, but we need this release
+          // function to be called before we can release the
+          // underlying frame. So, chain a promise into the release
+          // func so that the frame's release is blocked on this
+          // segment's release.
+          val p = new Promise[Unit]
+          val rest = copy(release = () => p.before(release()))
+          val releaseSegment = () => { p.setDone(); Future.Unit }
+          rest -> releaseSegment
+        } else {
+          // This frame has been entirely consumed, so return a
+          // release function that releases this frame (after all
+          // prior segments have been released).
+          val (rest, releaseRest) = tail.releasable()
+          val releaseFrame = () => release().before(releaseRest())
+          rest -> releaseFrame
+        }
```

Comment:
This should be changed.  Instead of blocking the final message's release future on prior futures, it should complete immediately -- but the last message to be released of a given frame should cause the flush...